### ISSUE TITLE
Error parsing code as Ruby 2.4 or 2.5

### ISSUE THREAD

**Issue by rrosenblum** [whitequark/parser#480]
The following ruby code produces a syntax error when using the Ruby 2.4 or Ruby 2.5 parser. I believe that this is related to #473. I was not able to reproduce the issue using `v2.5.0.3`.

`test.rb`
```ruby
puts "#{foo} #{(bar - baz)}"
bar.each do |qux|
  puts "#{qux}"
end
```

```
bundle exec ruby-parse --24 test.rb
test.rb:2:10: error: unexpected token kDO_BLOCK
test.rb:2: bar.each do |qux|
test.rb:2:   
```

**Comment by iliabylich** [comment]
@whitequark The error comes from the difference between the `parser` and MRI's grammar. On interpolation code lexer emits two tokens (and thus, calls surrounding logic for two tokens).

The first line of your code is not handled properly. Minimal case for reproducing: `m "#{}#{(foo)}"` (it's valid for parser, but `cmdarg` is `1` in the end of the line).

When lexer comes to the second interpolation it scans `#{`, emits tSTRING_BEG and tLPAREN:
```
m "#{}#{(foo)}"
     ^ tSTRING_DEND "}"                         interp_string  [0 <= cmdarg.1]
Shifting tSTRING_DEND }
pop_cmdarg
Lexer#emit tSTRING_DBEG #{
cmdarg.push(0)
Lexer#emit tLPAREN (
m "#{}#{(foo)}"
      ^~ tSTRING_DBEG "\#{"                     expr_beg      [10 <= cmdarg]
Shifting tSTRING_DBEG #{
push_cmdarg
m "#{}#{(foo)}"
        ^ tLPAREN "("                           expr_beg      [0 <= cmdarg.1]
Shifting tLPAREN (
```

Basically, instead of parentheses there can be array or hash brackets (`m "#{}#{[foo]}"` or `m "#{}#{{}}"`), because both instantly start a non-command context by doing `cmdarg.push(0)`.

MRI emits tokens one by one, that's the thing. That's why backported code doesn't work as it should.
Any ideas how to fix it?

**Comment by iliabylich** [comment]
@whitequark Changing the following action:
```
  action extend_interp_code {
    current_literal = literal
    current_literal.flush_string
    current_literal.extend_content

    emit(:tSTRING_DBEG, '#{'.freeze)

    if current_literal.heredoc?
      current_literal.saved_herebody_s = @herebody_s
      @herebody_s = nil
    end

    current_literal.start_interp_brace
    fcall expr_value;
  }
```
to have `fnext expr_value; fbreak;` instead of `fcall expr_value;`  should fix the issue, but `@cs` eventually becomes `nil` and the lexer dies. How can I trace why it happens?

**Comment by whitequark** [comment]
I can tell you. There's an
```
        fnext *stack_pop;
        fbreak;
```
somewhere, and that's what sets `@cs` to `nil`; it expects a counterpart `fcall`.

**Comment by maisumakun** [comment]
I have similar issue (only for Ruby 2.4 & 2.5 syntax in parser 2.5.0.5).

```ruby
class Test

  private def method(*args)
  end

  SOME_CONST = [].tap do |a|
    a.freeze
  end

end
```

produces an error like follows:

```
parse_bug.rb:9:23: error: unexpected token kDO_BLOCK
parse_bug.rb:9:   SOME_CONST = [].tap do |a|
parse_bug.rb:9:                       ^~
```

**Comment by whitequark** [comment]
That's a completely different issue.

**Comment by maisumakun** [comment]
@whitequark Thanks, I'll file a new issue.

**Comment by Mr0grog** [comment]
FWIW, I am also encountering this issue with some similar code (not *quite* the same). #482 seems to fix it for me, so it’s probably the same underlying cause:

```rb
func_call_without_parens func_with_interpolated_named_arg("..#{(x)}")

collection.each do |item|
end
```

…which outputs the same `unexpected token kDO_BLOCK` error.
